File: src/chapters/linear-systems/content/code/gauss-elimination.rs
// Naive Gaussian elimination (no pivoting) + back-substitution.
fn gauss_elimination(a: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = b.len();
    let mut m: Vec<Vec<f64>> = (0..n).map(|i| {
        let mut row = a[i].clone();
        row.push(b[i]);
        row
    }).collect();
    for k in 0..n {
        for i in k + 1..n {
            let f = m[i][k] / m[k][k];
            for j in k..=n { m[i][j] -= f * m[k][j]; }
        }
    }
    let mut x = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = m[i][n];
        for j in i + 1..n { s -= m[i][j] * x[j]; }
        x[i] = s / m[i][i];
    }
    x
}
fn main() {
    let a = vec![vec![2.0, 1.0, -1.0], vec![-3.0, -1.0, 2.0], vec![-2.0, 1.0, 2.0]];
    let b = vec![8.0, -11.0, -3.0];
    println!("{:?}", gauss_elimination(&a, &b));   // [2, 3, -1]
}

File: src/pages/practice/guides/matlab/gauss-elimination.m
function x = gauss_elimination(A, b)
% GAUSS_ELIMINATION  Naive Gaussian elimination (no pivoting) + back-substitution.
    n = numel(b); M = [A, b(:)];
    for k = 1:n-1
        for i = k+1:n
            f = M(i,k) / M(k,k);
            M(i, k:n+1) = M(i, k:n+1) - f * M(k, k:n+1);
        end
    end
    x = zeros(n, 1);
    for i = n:-1:1
        x(i) = (M(i, n+1) - M(i, i+1:n) * x(i+1:n)) / M(i, i);
    end
end

% --- Demo ---
A = [2 1 -1; -3 -1 2; -2 1 2]; b = [8; -11; -3];
disp(gauss_elimination(A, b)');   % 2 3 -1
